src/test/kotlin/com/terraformation/backend/TestEventPublisher.kt

```diff
@@ -10,67 +10,90 @@ class TestEventPublisher : ApplicationEventPublisher {
     publishedEvents.add(event)
   }
 
+  /**
+   * Clears the record of published events. This can be used in multi-step tests where an early step
+   * is supposed to publish an event but a later one isn't.
+   */
+  fun clear() {
+    publishedEvents.clear()
+  }
+
   /**
    * Asserts that a list of events were published in a specific order. Ignores any additional events
    * not on the list.
    */
-  fun assertEventsPublished(events: List<Any>) {
+  fun assertEventsPublished(events: List<Any>, message: String = "Expected events not published") {
     val expectedEvents = events.toSet()
     val publishedEventsFromExpectedList = publishedEvents.filter { it in expectedEvents }
 
-    assertEquals(events, publishedEventsFromExpectedList, "Expected events not published")
+    assertEquals(events, publishedEventsFromExpectedList, message)
   }
 
   /**
    * Asserts that a set of events were published in any order. Ignores any additional events not in
    * the set.
    */
-  fun assertEventsPublished(events: Set<Any>) {
+  fun assertEventsPublished(events: Set<Any>, message: String = "Expected events not published") {
     val publishedEventsFromExpectedSet = publishedEvents.filter { it in events }.toSet()
 
-    assertEquals(events, publishedEventsFromExpectedSet, "Expected events not published")
+    assertEquals(events, publishedEventsFromExpectedSet, message)
   }
 
   /**
    * Asserts that a list of events, and only that list of events, were published in a specific
    * order.
    */
-  fun assertExactEventsPublished(events: List<Any>) {
-    assertEquals(events, publishedEvents, "Expected events not published")
+  fun assertExactEventsPublished(
+      events: List<Any>,
+      message: String = "Expected events not published"
+  ) {
+    assertEquals(events, publishedEvents, message)
   }
 
   /** Asserts that a set of events, and only that set of events, were published in any order. */
-  fun assertExactEventsPublished(events: Set<Any>) {
-    assertEquals(events, publishedEvents.toSet(), "Expected events not published")
+  fun assertExactEventsPublished(
+      events: Set<Any>,
+      message: String = "Expected events not published"
+  ) {
+    assertEquals(events, publishedEvents.toSet(), message)
   }
 
   /** Asserts that a particular event was published. */
-  fun assertEventPublished(event: Any) {
+  fun assertEventPublished(event: Any, message: String = "Expected event not published") {
     if (event !in publishedEvents) {
       // Fail with an assertion that shows which events were actually published.
-      assertEquals(listOf(event), publishedEvents, "Expected event not published")
+      assertEquals(listOf(event), publishedEvents, message)
     }
   }
 
   /** Asserts that an event matching a predicate was published. */
-  fun assertEventPublished(predicate: (Any) -> Boolean) {
+  fun assertEventPublished(
+      message: String = "Expected event not published",
```

Comment:
just curious why message is the first arg here

Reply:
It's just to make the call sites a little nicer-looking. If the last argument of a function has a function type, you can call it with a trailing lambda like `assertEventPublished { ... }` or `assertEventPublished("message") { ... }`. If the function argument isn't the last one, you have to put the lambda in parentheses.